### DISCUSSION TITLE
Joining 2 lines of code into one

### DISCUSSION THREAD

**Discussion by tbmuc** [ImageMagick/ImageMagick#3015] [Help] [answered]
Hello,

I vould like to join these 2 lines into one
magick composite -gravity southwest -geometry +4685+750 %1-txt.png %1-tmp.jpg    %1-32-full.jpg
convert %1-32-full.jpg -background white -gravity Center  -extent 7056x4704 %1-32-poster.jpg

I tried these altenatives 
magick composite -gravity southwest -geometry +4685+750 %1-txt.png %1-tmp.jpg   %1-32-full.jpg -background white -gravity Center  -extent 7056x4704 %1-32-poster.jpg
and
magick ( composite -gravity southwest -geometry +4685+750 %1-txt.png %1-tmp.jpg   %1-32-full.jpg )  -background white -gravity Center  -extent 7056x4704 %1-32-poster.jpg

and got errors in both cases. 

How would the correct code looks like . IM7 Windows 10 bat

**Comment by snibgo** [comment]
First step: change "magick composite" to plain "magick". To do that, reverse the order of the input files, and insert "-composite" with the options before it:
```
magick %1-tmp.jpg %1-txt.png -gravity southwest -geometry +4685+750 -composite %1-32-full.jpg
```
Your second command is easy:
```
magick %1-32-full.jpg -background white -gravity Center -extent 7056x4704 %1-32-poster.jpg
```
The output of the first is the only input to the second, so merging them is easy:
```
magick %1-tmp.jpg %1-txt.png -gravity southwest -geometry +4685+750 -composite +write %1-32-full.jpg -background white -gravity Center -extent 7056x4704 %1-32-poster.jpg
```
For clarity, I would split this into lines. Windows BAT syntax:
```
magick ^
  %1-tmp.jpg %1-txt.png ^
  -gravity southwest -geometry +4685+750 -composite ^
  +write %1-32-full.jpg ^
  -background white -gravity Center -extent 7056x4704 ^
  %1-32-poster.jpg
```
If you don't want to save the intermediate result, you can move "+write %1-32-full.jpg".

**Comment by tbmuc** [comment]
Thank you very much for your support. snibgo.

I tried to follow your logic, but it seems I haven't understood completely. My end goal is, to speed up the script by removing big temporary files:

magick  %1-crop.jpg -gravity north -crop 2104:2995 +repage -resize 2104x2995!  %1-c32.jpg
magick  %TEMPLATE%  %1-c32.jpg -geometry +4798+1280   -composite %1-tmp.jpg
magick -size 4372x640 -font %FONT% -background None -fill #3b3734 caption:%2 -trim -resize 2336x320 -gravity center -extent 2336x320 %1-txt.png
magick %1-tmp.jpg %1-txt.png -gravity southwest -geometry +4685+750 -composite -background white -gravity Center -extent 7056x4704 %1-32-poster.jpg

As a second step, I tried to merge the two lines 

magick  %1-crop.jpg -gravity north -crop 2104:2995 +repage -resize 2104x2995!  %1-c32.jpg
magick  %TEMPLATE%  %1-c32.jpg -geometry +4798+1280   -composite %1-tmp.jpg

into 

magick (  %TEMPLATE%  %1-crop.jpg -gravity north -crop 2104:2995 +repage -resize 2104x2995!  ) -geometry +4798+1280   -composite %1-tmp.jpg

Surprisingly the content of the first line is missing,

**Comment by snibgo** [comment]
> -crop 2104:2995

This probably makes two images. The first image will be the largest chunk from the input that has your desired aspect ratio. The remainder of the input is put into a second image.

When you have two inputs, "-crop 2104:2995" will probably make four outputs.

When you want only one output for each input, use "-crop 2104:2995+0+0".

**Comment by tbmuc** [comment]
Ok, I tried it, but it has the same result:

magick  %TEMPLATE% ( %1-crop.jpg -gravity north -crop 2104:2995+0+0 +repage -resize 2104x2995! ) -geometry +4798+1280   -composite %1-tmp.jpg


 ( %1-crop.jpg -gravity north -crop 2104:2995+0+0 +repage -resize 2104x2995! )

seems not to appear on the compoiste image

**Comment by snibgo** [comment]
Please show sample input files for %TEMPLATE% and %1-crop.jpg.

**Comment by tbmuc** [comment]
%TEMPLATE% 
https://drive.google.com/file/d/1nNXwlNv4QHgw_wEzcwty6pZh2adG6q72/view?usp=sharing

%1-crop.jpg (the size differ, but the aspect ratio is close to that, however it neets a crop and a resize)
https://drive.google.com/file/d/19evbrWmqRs4-RA6gMM4qny17V5-QUb1c/view?usp=sharing

Please note: the origional script is working unless I try to merge lines to speed it up

**Comment by snibgo** [comment]
I suppose you want to replace the dog. You have "-geometry +4798+1280" which is the required position, measured from the top-left corner. But you also have "-gravity North", which isn't the top-left corner. So you need to insert "-gravity NorthWest", like this:
```
magick ih32tmp.jpg ( dog.jpg -gravity North -crop 2104:2995+0+0 +repage -resize "2104x2995^!" ) -gravity NorthWest -geometry +4798+1280 -composite out.jpg
```

**Comment by tbmuc** [comment]
Thank you very much. that was very helpful. May I ask 2 questions for a better understanding:

1) Is there a commad to reset all parameters to default values after creating a subimage.
2) The full statement below is much faster than the single steps but almost not readable. The commands are even nested. Are there something like subroutines in IM to structure it better. I would like to avoid file writing operations:

magick ( %TEMPLATE% ( %1-crop.jpg -gravity north -crop 2104:2995+0+0 +repage -resize 2104x2995! ) -gravity NorthWest -geometry +4798+1280   -composite ) ( -size 4372x640 -font %FONT% -background None -fill #3b3734 caption:%2 -trim -resize 2336x320 -gravity center -extent 2336x320 ) -gravity southwest -geometry +4685+750 -composite -background white -gravity Center -extent 7056x4704 %1-32-poster.jpg

**Comment by snibgo** [comment, marked as answer]
1. Reset parameters? Sadly, no.

2. Subroutines? Sadly, no. For readability, I split commands into multiple lines, where each line is a logical unit, like this (Windows BAT syntax):
```
magick ^
  ( %TEMPLATE% ^
    ( %1-crop.jpg -gravity north -crop 2104:2995+0+0 +repage -resize 2104x2995! ^
    ) ^
    -gravity NorthWest -geometry +4798+1280 -composite ^
  ) ^
  ( -size 4372x640 -font %FONT% -background None -fill #3b3734 caption:%2 ^
    -trim -resize 2336x320 ^
    -gravity center -extent 2336x320 ^
  ) ^
  -gravity southwest -geometry +4685+750 -composite ^
  -background white -gravity Center -extent 7056x4704 ^
  %1-32-poster.jpg
```
